### PR TITLE
AIP-38 Plug Graph on backend endpoint

### PR DESCRIPTION
**Pull request by pierrejeambrun**
This is just the rewiring. A follow up PR will take care of cleaning the types and simplify the interface. There are a lot of type casting, undefined / null check everywhere. I think we can simplify this.

Screenshot 2024-12-02 at 18 02 32
Screenshot 2024-12-02 at 18 02 53

### CODE REVIEW

File: airflow/api_fastapi/core_api/openapi/v1-generated.yaml [hunk 1]
```diff
@@ -7815,11 +7811,12 @@ components:
           - join
           - sensor
           - task
-          - task_group
+          - asset_condition
```
Comment (pierrejeambrun) [inline airflow/api_fastapi/core_api/openapi/v1-generated.yaml:7814]: Not used at the moment, but needed to remove an eslint error (non overlapping check on the enum value)

File: airflow/ui/src/pages/DagsList/Dag/Graph/Graph.tsx [hunk 2]
```diff
@@ -35,10 +36,16 @@ const nodeTypes = {
 };
 const edgeTypes = { custom: Edge };
 
-export const Graph = () => {
+export const Graph = ({ dagId }: { readonly dagId: DAGResponse["dag_id"] }) => {
   const { colorMode } = useColorMode();
 
   const { openGroupIds } = useOpenGroups();
+
+  const { data: graphData = { arrange: "LR", edges: [], nodes: [] } } =
```
Comment (pierrejeambrun) [inline airflow/ui/src/pages/DagsList/Dag/Graph/Graph.tsx:44]: Type system is not happy if I use `placeHolder` option. (Undefined is not type guarded, and fails later)

**Comment by tirkarthi** [comment]
Thanks for the PR, I was also working on the same thing yesterday but couldn't fix linters and typescript issues as a draft PR. Many of them were related to undefined vs null checks and the structure used in `./data` file being different compared to the API response causing type incompatibility.

https://github.com/tirkarthi/airflow/commit/e444d8459c70c1349271fa3b91d5d129480dc564

**Comment by tirkarthi** [comment]
Related issue for tracking https://github.com/apache/airflow/issues/44200

**Comment by tirkarthi** [comment]
I also remember seeing some errors in backend from pydantic where `rx` and `ry` were missing. Sample example dag example_setup_teardown had the issue.

**Comment by pierrejeambrun** [comment]
Yes, one of the issue we have is that I am not aware as of today of a way to have a datamodel generates something like:

```python
{
    field?: string
}
```

i.e have the field `optional/can be missing from the payload` (`?:`), but has no default value / not nullable...

**Comment by pierrejeambrun** [comment]
Both setup and teardown example dags (`example_setup_teardown` and `example_setup_teardown_taskflow`) seem to render properly:
Screenshot 2024-12-03 at 11 09 18
Screenshot 2024-12-03 at 11 09 36

File: airflow/ui/openapi-gen/requests/types.gen.ts [hunk 3]
```diff
@@ -761,15 +761,15 @@ export type JobResponse = {
...
   children?: Array<NodeResponse> | null;
-  id: string | null;
+  id: string;
   is_mapped?: boolean | null;
-  label?: string | null;
+  label: string;
   tooltip?: string | null;
   setup_teardown_type?: "setup" | "teardown" | null;
-  type: "join" | "sensor" | "task" | "task_group";
+  type: "join" | "sensor" | "task" | "asset_condition";
```
Comment (bbovenzi) [inline airflow/ui/openapi-gen/requests/types.gen.ts:769]: I just realized that we're missing the `operator`. On the task node type spec.

**Review by bbovenzi — APPROVED** [review]
lgtm. I will give a pass at trying to improve the typing in another PR.

**Pull request merged**